pkg/controller/clusterstate/clusterstate_controller.go

```diff
@@ -0,0 +1,303 @@
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Info("cluster deployment not found")
+			return reconcile.Result{}, nil
+		}
+		logger.WithError(err).Error("Error getting cluster deployment")
+		return reconcile.Result{}, err
+	}
+	if !cd.DeletionTimestamp.IsZero() {
+		logger.Infof("ClusterDeployment resource has been deleted")
+		return reconcile.Result{}, nil
+	}
+	if !cd.Status.Installed || len(cd.Status.AdminKubeconfigSecret.Name) == 0 {
+		logger.Infof("ClusterDeployment is not ready")
+		return reconcile.Result{}, nil
+	}
+	kubeconfigSecret := &corev1.Secret{}
+	err = r.Get(context.Background(), types.NamespacedName{Namespace: cd.Namespace, Name: cd.Status.AdminKubeconfigSecret.Name}, kubeconfigSecret)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Warn("admin kubeconfig does not yet exist")
+		} else {
+			return reconcile.Result{}, err
+		}
+	}
+	kubeconfig, err := controllerutils.FixupKubeconfigSecretData(kubeconfigSecret.Data)
+	remoteClient, err := r.remoteClientBuilder(string(kubeconfig), controllerName)
+	if err != nil {
+		logger.WithError(err).Error("error building remote cluster client connection")
+		return reconcile.Result{}, err
+	}
+
+	clusterOperators := &configv1.ClusterOperatorList{}
+	err = remoteClient.List(context.TODO(), clusterOperators)
+	if err != nil {
+		logger.WithError(err).Errorf("failed to list target cluster operators")
+		return reconcile.Result{}, err
+	}
+
+	operatorStates := make([]hivev1.ClusterOperatorState, 0, len(clusterOperators.Items))
+	for _, clusterOperator := range clusterOperators.Items {
+		operatorStates = append(operatorStates, hivev1.ClusterOperatorState{
+			Name:       clusterOperator.Name,
+			Conditions: clusterOperator.Status.Conditions,
+		})
+	}
+	if operatorStatesChanged(logger, st.Status.ClusterOperators, operatorStates) {
+		st.Status.ClusterOperators = operatorStates
+		err = r.updateStatus(r, st)
+		if err != nil {
+			logger.WithError(err).Errorf("failed to update cluster operator state")
+			return reconcile.Result{}, err
+		}
+		logger.Info("clusterState has been updated")
+		return reconcile.Result{}, nil
+	}
+	return reconcile.Result{
+		RequeueAfter: statusUpdateInterval,
+	}, nil
+}
+
+func operatorStatesChanged(logger log.FieldLogger, existing, updated []hivev1.ClusterOperatorState) bool {
+	changed := false
+	existingNames := sets.NewString()
+	updatedNames := sets.NewString()
+
+	for _, i := range existing {
+		existingNames.Insert(i.Name)
+	}
+	for _, i := range updated {
+		updatedNames.Insert(i.Name)
+	}
+
+	removed := existingNames.Difference(updatedNames)
+	added := updatedNames.Difference(existingNames)
+	same := existingNames.Intersection(updatedNames)
+	if removed.Len() > 0 {
+		changed = true
+		logger.Infof("Removed cluster operators: %v", removed.List())
+	}
+	if added.Len() > 0 {
+		changed = true
+		logger.Infof("Added cluster operators: %v", added.List())
+	}
+
+	for _, name := range same.List() {
```

Comment:
Yes that was the purpose. My thinking in having all this code here is that it could be valuable for debugging purposes to understand when a particular operator change was first observed on a cluster.

Reply:
OK. I can live with that.